### DISCUSSION TITLE
The type 'Application' is defined in an assembly that is not referenced. You must add a reference to assembly 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'

### DISCUSSION THREAD

**Discussion by Mike-E-angelo** [unoplatform/uno#19066] [Q&A] [answered]
On building for the `net9.0-windows10.0.26100` target framework, I am getting the following error:

```
The type 'Application' is defined in an assembly that is not referenced. You must add a reference to assembly 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'.
```

The only reference I have is in a csproj that looks like the following:

```
<Project Sdk="Uno.Sdk">
	<ItemGroup>
		<ProjectReference Include="..\DragonSpark.Application\DragonSpark.Application.csproj" />
	</ItemGroup>
	<ItemGroup>
		<PackageReference Include="Uno.Extensions.Hosting.WinUI" />
	</ItemGroup>
</Project>
```

I found this StackOverflow question discussing the error:
https://stackoverflow.com/questions/70034567/type-universe-cannot-resolve-assembly-uno-ui-version-255-255-255-255

But it does not appear to apply to me as I am not using the toolkit (that I know of).

Any assistance in determining this error and how to solve it would be appreciated.  Thank you for your assistance.

**Comment by jeromelaban** [comment]
Thanks for the report!

Are you able to extract a repro?

**Reply by Mike-E-angelo** [reply]
Thank you for your investigation @jeromelaban 🙏  I was able to do so here:
https://github.com/DragonSpark/Framework/blob/issues/reported/uno/19066/The%20DragonSpark%20Framework.sln

The error occurs here:
https://github.com/DragonSpark/Framework/blob/90602b53d06ff1e6450bb3781cd95dfe07911376/UnoApp1/UnoApp1/App.xaml.cs#L21-L21

Please let me know if any further information is required and I will assist you. 👍

**Reply by jeromelaban** [reply, marked as answer]
Thanks. The issue here is that the projects tree from [this point](https://github.com/DragonSpark/Framework/blob/90602b53d06ff1e6450bb3781cd95dfe07911376/UnoApp1/UnoApp1/UnoApp1.csproj#L51), only has one `net9.0` target framework which is used by uno to refer to skia/webassembly compatible targets.

To fix this, you an either specify a list of `TargetFrameworks` in all the projects referenced by your head that matches at least your top-level project (if you want to use Uno API in those projects), or you can replace the `Sdk="uno.sdk"` line by `Sdk="Microsoft.Net.Sdk"` at the top of the child projects (if you only want to use plain .NET BCL APIs).

The Uno Platform Library template contains the `TargetFrameworks` that need to be present, if you need an example.

Note also that it is generally not a good idea to set the `TargetFramework` property in `Directory.Build.props`, as it will override any project  that uses `TargetFrameworks` and prevent proper multi-targeting from working.

**Reply by Mike-E-angelo** [reply]
Thank you for the quick answer @jeromelaban it is greatly appreciated. 🙏

>  or you can replace the Sdk="uno.sdk" line by Sdk="Microsoft.Net.Sdk" at the top of the child projects

So believe it or not this is one of the first things I tried and it still resulted in the error.  But I have [noticed caching errors with Visual Studio](https://developercommunity.visualstudio.com/t/ALT-B-U-build-project-Does-not-Save-Fi/10599589) when building so maybe I was running into that.  I will see if I can get this to work.

> Note also that it is generally not a good idea to set the TargetFramework property in Directory.Build.props, as it will override any project that uses TargetFrameworks and prevent proper multi-targeting from working.

That's a good point!  I will keep that in mind going forward.  Note that the Framework currently uses this but the solution that I am working on (not in GitHub) is not currently structured this way.

Thank you again for all your assistance and for all your efforts out there. 👍

**Comment by DevTKSS** [comment]
@Mike-E-angelo I think mine done the same in the past. I normally do a "cleanup projectmap" and rebuild it. If that not helps at this point, reopen visual studio, to cause the nuget restore packages normally resolves it on my side 😄

**Reply by Mike-E-angelo** [reply]
Thank you for the suggestion @DevTKSS I did try this and it appears that this is another issue.  Well, I should say I tried restarting Visual Studio and that did not work.  Can you specify what you mean by `cleanup projectmap` ?  I am a total newb with the Uno ecosystem. 😁

**Reply by DevTKSS** [reply]
Thats just visual studio general. Above the debug button you can run build, cleanup and "neu erstellen" think will be translated as rebuild or restore, should ne right below. The actions can be runned in different scopes like project or the whole project map. Thats essentially cleaning up the bin and obj folder which could contain generator code where the generator got confused and caused this issues. Had this kind of issues today with a vb.net winforms project at work which suddenly thought to forget about the existence of mscorlib.dll and throwed me a "system void is not recognized" statement 🤣 not getting boring. Thats not at all platform specific, just the thing i do on my projects when such strange things do happen.
